runtime/gc_modron_startup/mmparse.cpp

```diff
@@ -559,18 +559,127 @@ xlpSubOptionsParser(J9JavaVM *vm, IDATA xlpIndex, XlpError *xlpError, UDATA *req
+	if (!isRequestedSizeSupported) {
+		/* Print a message indicating requested page size is not supported and a different page size will be used */
+		const char *oldQualifier, *newQualifier;
+		const char *oldPageType = NULL;
+		const char *newPageType = NULL;
+		UDATA oldSize = requestedPageSize;
+		UDATA newSize = pageSize;
+		qualifiedSize(&oldSize, &oldQualifier);
+		qualifiedSize(&newSize, &newQualifier);
+		if (0 == (J9PORT_VMEM_PAGE_FLAG_NOT_USED & requestedPageFlags)) {
+			oldPageType = getPageTypeString(requestedPageFlags);
+		}
+		if (0 == (J9PORT_VMEM_PAGE_FLAG_NOT_USED & pageFlags)) {
+			newPageType = getPageTypeString(pageFlags);
+		}
+		if (NULL == oldPageType) {
+			if (NULL == newPageType) {
+				j9nls_printf(PORTLIB, J9NLS_INFO, J9NLS_GC_OPTIONS_LARGE_PAGE_SIZE_NOT_SUPPORTED, oldSize, oldQualifier, newSize, newQualifier);
+			} else {
+				j9nls_printf(PORTLIB, J9NLS_INFO, J9NLS_GC_OPTIONS_LARGE_PAGE_SIZE_NOT_SUPPORTED_WITH_NEW_PAGETYPE, oldSize, oldQualifier, newSize, newQualifier, newPageType);
+			}
+		} else {
+			if (NULL == newPageType) {
+				j9nls_printf(PORTLIB, J9NLS_INFO, J9NLS_GC_OPTIONS_LARGE_PAGE_SIZE_NOT_SUPPORTED_WITH_REQUESTED_PAGETYPE, oldSize, oldQualifier, oldPageType, newSize, newQualifier);
+			} else {
+				j9nls_printf(PORTLIB, J9NLS_INFO, J9NLS_GC_OPTIONS_LARGE_PAGE_SIZE_NOT_SUPPORTED_WITH_PAGETYPE, oldSize, oldQualifier, oldPageType, newSize, newQualifier, newPageType);
+			}
+		}
+	}
+}
+
+/**
+ * 
+ * Obtains and sets the largest page size available on the system. 
+ * Returns true if the operation was successful.
+ * 
+ */
+bool
+gcSetPreferredPageSize(J9JavaVM *vm, MM_GCExtensions *extensions)
+{
+	UDATA defaultLargePageSize = 0;
+	UDATA defaultLargePageFlags = J9PORT_VMEM_PAGE_FLAG_NOT_USED;
+	PORT_ACCESS_FROM_JAVAVM(vm);
+
+	j9vmem_default_large_page_size_ex(0, &defaultLargePageSize, &defaultLargePageFlags);
+	
+	if (0 != defaultLargePageSize) {
+		extensions->requestedPageSize = defaultLargePageSize;
+		extensions->requestedPageFlags = defaultLargePageFlags;
+	} else {
+		return false;
+	}
+
+	return true;
+}
+
+/**
+ * 
+ *  Sets the large page size used within the objectheap when the options -XX:+UseLargePages or -XX:LargePageSizeInBytes=<size> are used.
+ *  The value of the large page size will only be overwritten when these arguments are located in the right most index.
+ *  This is compared to the indices of -Xlp, -Xlp<size>, and -Xlp:objectheap:pagesize=<xx>/
+ * 
+ */
+static bool
+gcProcessXXObjectHeapArguments(J9JavaVM *vm)
+{
+	PORT_ACCESS_FROM_JAVAVM(vm);
+	MM_GCExtensions *extensions = MM_GCExtensions::getExtensions(vm);
+	if (vm->largePageSizeRequested == -1) {
+		if (!gcSetPreferredPageSize(vm, extensions)) {
+			j9nls_printf(PORTLIB, J9NLS_ERROR, J9NLS_GC_OPTIONS_SYSTEM_CONFIG_OPTION_NOT_SUPPORTED, "-XX:+UseLargePages");
+			return false;
+		}
+	} else {
+		UDATA largePageSizeRequested = vm->largePageSizeRequested;
+		UDATA largePageFlagRequested = J9PORT_VMEM_PAGE_FLAG_NOT_USED;
```

Comment:
This would not get us desired behaviour on ZOS. `omrvmem_find_valid_page_size()` would force to use `OMRPORT_VMEM_PAGE_FLAG_FIXED` however we need to try `OMRPORT_VMEM_PAGE_FLAG_PAGEABLE` first. 
https://github.com/eclipse/omr/blob/88f4df72d15347a8b08f6cb16edb331a59978443/port/zos390/omrvmem.c#L1359

Reply:
There is another ZOS specific detail we should keep in mind: `omrvmem_find_valid_page_size()` would not be able to find default small page size as far as it does search in available pages array from index 1 (not 0):
https://github.com/eclipse/omr/blob/88f4df72d15347a8b08f6cb16edb331a59978443/port/zos390/omrvmem.c#L1561
This does not mean your code change existing behaviour but you should be aware